### ISSUE TITLE
I can't add a new license and rule

### ISSUE THREAD

**Issue by MihaiChirciu** [nexB/scancode-toolkit#3401]
Hi

I am new to Scancode and I did not manage to add a new license or rule. I have installed Scancode (version 31.2.4) on a windows 10 machine, via Git Bash by cloning the github project. I ran the tool on some test files and it seems to be working fine, however I wanted add a new license and a new rule for it.

My files: [zz-test.zip](https://github.com/nexB/scancode-toolkit/files/11499750/zz-test.zip)
I have created zz-test.LICENSE and 4 .RULE files (zz-test.RULE and zz-test_1 to 3). I created the license file here: scancode-toolkit\src\licensedcode\data\licenses and the rule files here: scancode-toolkit\src\licensedcode\data\rules

I have attached the files. After creating the file I ran "scancode --reindex-licenses" and then I ran "scancode -clpeui --html test_files_results.html test_files". The "test_files" directory contains only one file: mihai_file.h that contains:
/*
 *  Copyright Mihai C.
 *
 *  This is my custom test license.
 *  Use for testing purposes only!
 */

#ifndef _SAMPLE_DRV_H_
#define _SAMPLE_DRV_H_

I used to read the documentation from https://scancode-toolkit.readthedocs.io/en/stable/how-to-guides/add_new_license_detection_rule.html but since yesterday I also read the https://scancode-toolkit.readthedocs.io/en/latest/how-to-guides/add_new_license_detection_rule.html

Sadly, I still cannot figure what I am doing wrong or why the tool does not detect my test license.

**Comment by DennisClark** [comment]
@AyanSinhaMahapatra    please take a look at this problem when you can get to it, or re-assign to someone who understands how the rules work, thanks.

**Comment by AyanSinhaMahapatra** [comment]
@DennisClark oops, this one slipped through the cracks :sweat_smile: Looking into this now.

**Comment by AyanSinhaMahapatra** [comment]
@MihaiChirciu sorry for the really late reply. 

I tried to reproduce the issue with your examples. Couple of points:

1. If you did a fresh clone of the `scancode-toolkit` repo and then added just your license file `zz-test.LICENSE` such that it's file path was at `scancode-toolkit/src/licensedcode/data/licenses/zz-test.LICENSE` and then did a installation https://scancode-toolkit.readthedocs.io/en/latest/getting-started/install.html#installation-from-source-code-git-clone, and thereafter activated the virtualenv and run `scancode-reindex-licenses` you would have gotten the following error:
```
licensedcode.models.InvalidLicense: Errors while validating licenses:

License: zz-test
  'No category: Use "Unstated License" if not known.'
  'No SPDX license key'
```
which means you had to add the following attributes to your license file:
```
category: Permissive
spdx_license_key: Scancode-licenseref-zz-test
```
Then the reindex command would pass, which means your license addition has been added to the index.

2. All your rules have their license expression set to the name of the `zz-test.LICENSE` license: `name: Mihai License`
If you run the reindex command after adding one of this rules, you will get the following error:
```
licensedcode.models.InvalidRule: Errors while validating rules:

Invalid license_expression: Mihai License ,keys should be lowercase.
  Rule(identifier='zz-test.RULE', license_expression='Mihai License', minimum_coverage=80, is_continuous=False, relevance=100, has_stored_relevance=False, length=0)
    file:///home/ayansinha/nexB/write_access/scancode-extra/src/licensedcode/data/rules/zz-test.RULE

Failed to parse and validate license_expression: Mihai License with error: Invalid symbols sequence such as (A B) for token: "License" at position: 6
  Rule(identifier='zz-test.RULE', license_expression='Mihai License', minimum_coverage=80, is_continuous=False, relevance=100, has_stored_relevance=False, length=0)
    file:///home/ayansinha/nexB/write_access/scancode-extra/src/licensedcode/data/rules/zz-test.RULE

```
The rules all have:
 `license_expression: Mihai License`
Instead they should be:
`license_expression: zz-test`

As license expressions are made from the combination of license keys, not license names.

3. After making the above changes, if you add all the files, the license and the keys and run the reindex command you will get the following error:
```
licensedcode.index.DuplicateRuleError: Duplicate rules:
zz-test.RULE
zz-test_1.RULE
```
Whitespace and special characters are ignored when creating the index, so your two rules are essentially the same here. So you can only have on of them.

4. After making the above mentioned changes, if you run the `scancode-reindex-licenses` it will pass, meaning your license and rules are added to the index successfully. Then if you run the following command:
`scancode --license --license-text mihai_file.h --json-pp mihai.json`
You will get the following result.json

```
{
  "headers": [
    {
      "tool_name": "scancode-toolkit",
      "tool_version": "v32.0.5rc3",
      "options": {
        "input": [
          "mihai_file.h"
        ],
        "--json-pp": "mihai.json",
        "--license": true,
        "--license-text": true
      },
      "notice": "Generated with ScanCode and provided on an \"AS IS\" BASIS, WITHOUT WARRANTIES\nOR CONDITIONS OF ANY KIND, either express or implied. No content created from\nScanCode should be considered or used as legal advice. Consult an Attorney\nfor any legal advice.\nScanCode is a free software code scanning tool from nexB Inc. and others.\nVisit https://github.com/nexB/scancode-toolkit/ for support and download.",
      "start_timestamp": "2023-06-23T204822.230115",
      "end_timestamp": "2023-06-23T204823.563404",
      "output_format_version": "3.0.0",
      "duration": 1.333315372467041,
      "message": null,
      "errors": [],
      "warnings": [],
      "extra_data": {
        "system_environment": {
          "operating_system": "linux",
          "cpu_architecture": "64",
          "platform": "Linux-5.15.0-75-generic-x86_64-with-glibc2.29",
          "platform_version": "#82~20.04.1-Ubuntu SMP Wed Jun 7 19:37:37 UTC 2023",
          "python_version": "3.8.10 (default, May 26 2023, 14:05:08) \n[GCC 9.4.0]"
        },
        "spdx_license_list_version": "3.20",
        "files_count": 1
      }
    }
  ],
  "license_detections": [
    {
      "identifier": "zz_test-0daf9ff7-f96f-f3e4-65e2-9cb303ae96b3",
      "license_expression": "zz-test",
      "detection_count": 1
    }
  ],
  "files": [
    {
      "path": "mihai_file.h",
      "type": "file",
      "detected_license_expression": "zz-test",
      "detected_license_expression_spdx": "Scancode-licenseref-zz-test",
      "license_detections": [
        {
          "license_expression": "zz-test",
          "matches": [
            {
              "score": 61.0,
              "start_line": 4,
              "end_line": 5,
              "matched_length": 11,
              "match_coverage": 100.0,
              "matcher": "2-aho",
              "license_expression": "zz-test",
              "rule_identifier": "zz-test.RULE",
              "rule_relevance": 61,
              "rule_url": "https://github.com/nexB/scancode-toolkit/tree/develop/src/licensedcode/data/rules/zz-test.RULE",
              "matched_text": "    This is my custom test license.\n    Use for testing purposes only!"
            }
          ],
          "identifier": "zz_test-0daf9ff7-f96f-f3e4-65e2-9cb303ae96b3"
        }
      ],
      "license_clues": [],
      "percentage_of_license_text": 50.0,
      "scan_errors": []
    }
  ]
}
```
And your example license in your test file will be detected correctly.

**Comment by MihaiChirciu** [comment]
@AyanSinhaMahapatra Hi and thank you so much for your help. It is working now.

I had my files changed after I submitted this case, based on the already existing licenses and rules. But it still wouldn't work for me.

I also identified what I was doing wrong. I was running:
$ scancode --reindex-licenses
Rebuilding the license index...
Done.
Instead of scancode-reindex-licenses. So everytime I was assuming that the reindex was successfully since I saw no errors.
